frame/support/src/storage/transactional.rs

```diff
@@ -158,6 +158,40 @@ pub fn with_transaction_unchecked<R>(f: impl FnOnce() -> TransactionOutcome<R>)
 	}
 }
 
+/// Execute the supplied function, adding a new storage layer.
+///
+/// This is the same as `with_transaction`, but assuming that any function returning
+/// an `Err` should rollback, and any function returning `Ok` should commit. This
+/// provides a cleaner API to the developer who wants this behavior.
+pub fn with_storage_layer<T, E>(f: impl FnOnce() -> Result<T, E>) -> Result<T, E>
+where
+	E: From<DispatchError>,
```

Comment:
No, it needs to be `From` because `with_transaction` has this line:

```
	let _guard = inc_transaction_level().map_err(|()| TransactionalError::LimitReached.into())?;
```

Reply:
Huh, why does that line require an `into()`? `DispatchError` already has a `From<TransactionalError>` implementation, and hence `Into<DispatchError>` is automatically implemented for `TransactionalError`, i.e. leaving the error type as `TransactionalError` should work as-is if the return error type is `Into<DispatchError>`.

Reply:
Okay nevermind, the `?` operator only performs automatic error type conversions on `From` trait impls, not `Into` trait impls.